xcm/xcm-builder/src/universal_exports.rs

```diff
@@ -0,0 +1,368 @@
+		// export for free. Common-good chains will typically be afforded this.
+		let message =
+			Xcm(vec![ExportMessage { network: remote_network, destination: remote_location, xcm }]);
+		let (v, mut cost) = validate_send::<Router>(bridge, message)?;
+		if let Some(payment) = maybe_payment {
+			cost.push(payment);
+		}
+		Ok((v, cost))
+	}
+
+	fn deliver(validation: Router::Ticket) -> Result<XcmHash, SendError> {
+		Router::deliver(validation)
+	}
+}
+
+/// Implementation of `SendXcm` which wraps the message inside an `ExportMessage` instruction
+/// and sends it to a destination known to be able to handle it.
+///
+/// The `ExportMessage` instruction on the bridge is paid for from the local chain's sovereign
+/// account on the bridge. The amount paid is determined through the `ExporterFor` trait.
+pub struct SovereignPaidRemoteExporter<Bridges, Router, UniversalLocation>(
+	PhantomData<(Bridges, Router, UniversalLocation)>,
+);
+impl<Bridges: ExporterFor, Router: SendXcm, UniversalLocation: Get<InteriorMultiLocation>> SendXcm
+	for SovereignPaidRemoteExporter<Bridges, Router, UniversalLocation>
+{
+	type Ticket = Router::Ticket;
+
+	fn validate(
+		dest: &mut Option<MultiLocation>,
+		xcm: &mut Option<Xcm<()>>,
+	) -> SendResult<Router::Ticket> {
+		let d = dest.as_ref().ok_or(MissingArgument)?.clone();
+		let devolved = ensure_is_remote(UniversalLocation::get(), d).map_err(|_| NotApplicable)?;
+		let (remote_network, remote_location) = devolved;
+
+		let xcm = xcm.take().ok_or(MissingArgument)?;
+		let (bridge, maybe_payment) =
+			Bridges::exporter_for(&remote_network, &remote_location, &xcm).ok_or(NotApplicable)?;
+
+		let local_from_bridge =
+			UniversalLocation::get().invert_target(&bridge).map_err(|_| Unroutable)?;
+		let export_instruction =
+			ExportMessage { network: remote_network, destination: remote_location, xcm };
+
+		let message = Xcm(if let Some(ref payment) = maybe_payment {
+			let fees = payment
+				.clone()
+				.reanchored(&bridge, UniversalLocation::get())
+				.map_err(|_| Unroutable)?;
+			vec![
+				WithdrawAsset(fees.clone().into()),
+				BuyExecution { fees, weight_limit: Unlimited },
+				export_instruction,
+				RefundSurplus,
+				DepositAsset { assets: All.into(), beneficiary: local_from_bridge },
+			]
+		} else {
+			vec![export_instruction]
+		});
+
+		// We then send a normal message to the bridge asking it to export the prepended
+		// message to the remote chain. This will only work if the bridge will do the message
+		// export for free. Common-good chains will typically be afforded this.
+		let (v, mut cost) = validate_send::<Router>(bridge, message)?;
+		if let Some(bridge_payment) = maybe_payment {
+			cost.push(bridge_payment);
+		}
+		Ok((v, cost))
+	}
+
+	fn deliver(ticket: Router::Ticket) -> Result<XcmHash, SendError> {
+		Router::deliver(ticket)
+	}
+}
+
+pub trait DispatchBlob {
+	/// Dispatches an incoming blob and returns the unexpectable weight consumed by the dispatch.
```

Comment:
what does it mean `returns the unexpectable weight`?

It returns `Result<()` and also few lines bellow, there is unused result from `send_xcm` which is `Result<(XcmHash, MultiAssets)`
```
		...
		send_xcm::<Router>(dest, message).map_err(|_| DispatchBlobError::RoutingError)?;
		Ok(())
		...
```

Could be `unexpectable weight` related to that send_xcm's MultiAssets?

Reply:
nit: here is still misleading doc `and returns the unexpectable weight consumed by the dispatch` vs `Result<()`

Reply:
Unexpectable weight is basically the weight which the dispatchable actually used. It's named as such because it's depednent on the parameter and therefore not benchmarkable.

Reply:
@gavofyork 
thank you very much, but I am still little bit confused, because `fn dispatch_blob(blob: Vec<u8>) -> Result<(), DispatchBlobError>`, returns just `()` and not any `Weight / Option<Weight>` or something similar.

How should the implementation of `fn dispatch_blob(..) -> Result<(), ` return `Unexpectable weight that was consumed`?

Reply:
The docs are wrong.

Reply:
Fixed in https://github.com/paritytech/polkadot/pull/7230